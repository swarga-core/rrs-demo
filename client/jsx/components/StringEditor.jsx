import React, { Component } from 'react';
import PropTypes from 'prop-types';
import isEqual from 'lodash/isEqual';


class StringEditor extends Component {

  componentDidMount() {
    if (this.editor) {
      this.editor.focus();
    }
  }

  componentDidUpdate() {
    if (this.editor) {
      this.editor.focus();
    }
  }

  shouldComponentUpdate(nextProps) {
    return (
      !isEqual(nextProps.item, this.props.item)
      || nextProps.isEditing !== this.props.isEditing
    );
  }

  onEditorKeyDown = (event) => {
    switch(event.keyCode) {
      case 13:
        if (this.editor.value) {
          this.props.onEditorUpdate(this.editor.value);
        }
        break;
      case 27:
        this.props.onEditorCancel();
        break;
    }
  }

  onBlur = () => {
    this.props.onEditorCancel();
  }

  render() {
    const { item, fieldName, isEditing } = this.props;
    if (isEditing) {
      return (
        <input
          type="text"
          className="form-control input-sm"
          defaultValue={item[fieldName]}
          ref={editor => this.editor = editor}
          onKeyDown={this.onEditorKeyDown}
          onBlur={this.onBlur}
        />
      );
    } else {
      return (<div>{item[fieldName]}</div>);
    }
  }

}

StringEditor.propTypes = {
  fieldName: PropTypes.string.isRequired,
  item: PropTypes.object.isRequired,
  state: PropTypes.object.isRequired,
  schema: PropTypes.object.isRequired,
  isEditing: PropTypes.bool.isRequired,
  onEditorUpdate: PropTypes.func.isRequired,
  onEditorCancel: PropTypes.func.isRequired,
};

export default StringEditor;