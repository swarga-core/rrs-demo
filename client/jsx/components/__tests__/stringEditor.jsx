import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import StringEditor from '../StringEditor';

const defaultNameValue = 'Test Department';
const defaultProps = {
  item: {id: 1, name: defaultNameValue},
  schema: {},
  state: {},
  isEditing: false,
  fieldName: 'name',
  onEditorUpdate: ()=>{},
  onEditorCancel: ()=>{}
};

describe('<String Editor />', ()=>{

  describe('in text mode', () => {

    it('has matched snapshot', ()=>{
      const component = renderer.create(
        <StringEditor {...defaultProps} />
      );
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('has right value', ()=>{
      const editor = shallow(
        <StringEditor {...defaultProps} />);
      expect(editor.text()).toEqual(defaultNameValue);
    });

  });

  describe('in edit mode', () => {

    const editModeProps = { ...defaultProps, isEditing: true };

    it('has matched snapshot', ()=>{
      const component = renderer.create(
        <StringEditor {...editModeProps} />
      );
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('calls props.actions.onEditorCancel() after pressing "Escape"', ()=>{
      const mockOnEditorCancel = jest.fn();
      const props = { ...editModeProps, onEditorCancel: mockOnEditorCancel };
      const editor = shallow(<StringEditor { ...props } />);
      editor.simulate('keydown', { keyCode: 27 });
      expect(mockOnEditorCancel.mock.calls.length).toBe(1);
    });

    it('calls props.actions.onEditorCancel() on blur event', ()=>{
      const mockOnEditorCancel = jest.fn();
      const props = { ...editModeProps, onEditorCancel: mockOnEditorCancel };
      const editor = shallow(<StringEditor { ...props } />);
      editor.find('input').simulate('blur');
      expect(mockOnEditorCancel.mock.calls.length).toBe(1);
    });

  });

});
