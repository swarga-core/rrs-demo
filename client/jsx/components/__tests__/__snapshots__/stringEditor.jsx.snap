// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<String Editor /> in edit mode has matched snapshot 1`] = `
<input
  className="form-control input-sm"
  defaultValue="Test Department"
  onBlur={[Function]}
  onKeyDown={[Function]}
  type="text"
/>
`;

exports[`<String Editor /> in text mode has matched snapshot 1`] = `
<div>
  Test Department
</div>
`;
